pkg/operator/ceph/object/spec.go

```diff
@@ -912,3 +922,70 @@ func renderProbe(cfg rgwProbeConfig) (string, error) {
 
 	return writer.String(), nil
 }
+
+func (c *clusterConfig) addDNSNamesToRGWServer() (string, error) {
+	if (c.store.Spec.Hosting == nil) || len(c.store.Spec.Hosting.DNSNames) <= 0 {
+		return "", nil
+	}
+	if !c.clusterInfo.CephVersion.IsAtLeastReef() {
+		return "", errors.New("rgw dns names are supported from ceph v18 onwards")
+	}
+
+	// add default RGW service name to dns names
+	dnsNames := c.store.Spec.Hosting.DNSNames
+	dnsNames = append(dnsNames, domainNameOfService(c.store))
+
+	// add custom endpoints from zone spec if exists
+	if c.store.Spec.Zone.Name != "" {
+		zone, err := c.context.RookClientset.CephV1().CephObjectZones(c.store.Namespace).Get(c.clusterInfo.Context, c.store.Spec.Zone.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		for _, endpoint := range zone.Spec.CustomEndpoints {
+			domainName, err := URLParse(endpoint)
+			if err != nil {
+				return "", err
+			}
+			if domainName != "" {
+				dnsNames = append(dnsNames, domainName)
+			}
+		}
+	}
+
+	// remove duplicate dns names
+	checkDuplicate := make(map[string]bool)
+	removeDuplicateDNSNames := []string{}
+	for _, dnsName := range dnsNames {
+		if _, ok := checkDuplicate[dnsName]; !ok {
+			checkDuplicate[dnsName] = true
+			removeDuplicateDNSNames = append(removeDuplicateDNSNames, dnsName)
+		}
+	}
+
+	err := validateDNS1123(removeDuplicateDNSNames)
+	if err != nil {
+		return "", err
+	}
+
+	return cephconfig.NewFlag("rgw dns name", strings.Join(removeDuplicateDNSNames, ",")), nil
+}
+
+// Validate DNSName if it is codified by RFC-1123
+func validateDNS1123(dnsNames []string) error {
+	var err []string
+	for _, i := range dnsNames {
+		err = validation.IsDNS1123Subdomain(i)
+		if err != nil {
+			return errors.Errorf(strings.Join(err, " "), "invalid DNS-1123 subdomain %q", i)
+		}
+	}
+	return nil
+}
+
+func URLParse(endpoint string) (string, error) {
+	parsedURL, err := url.Parse(endpoint)
+	if err != nil {
+		return "", err
+	}
+	return parsedURL.Hostname(), nil
+}
```

Comment:
Played with this here after your last slack message: https://play.golang.com/p/0DPQXbbzaU6

Copying relevant parts here. 
```
	fmt.Println(URLParse("http://my.host:80"))
	// my.host <nil>
	// ^ HTTP happy path OK
	
	fmt.Println(URLParse("https://my.host:80"))
	// my.host <nil>
	// ^ HTTPS happy path OK

	fmt.Println(URLParse("httt://my.host:80"))
	// my.host <nil>
	// ^ I think this is expected from this lib because it would have to handle different protocols as `<protocol>://` 
	// like `fttp://`, `file://`, `customproto://`, etc.
	// Would anyone have a reason to use `s3://` as an endpoint URL? I'm not sure.
	// It's **probably** safe, and **possibly** even best for us to allow this scenario, 
	//   but IMO we should still see what Casey and Kyle Bader think.
	
	fmt.Println(URLParse("http://my.host:80:8080"))
	// my.host:80 <nil>
	// ^ I think this is NOT okay though. DNS spec doesn't allow `:` in names, so this should be a failure IMO.
	
	fmt.Println(URLParse("192.168.1.1:80"))
	// "" parse "192.168.1.1:80": first path segment in URL cannot contain colon
	// ^ We should check to see what Ceph allows here. 
	// IIUC, this is the behavior we want here because we are supposed to be giving ceph a hostname as input, 
	//   and an IP address is not the same as a hostname.
	// I think this boils down to whether RGW allows `<bucketname>.192.168.1.1` wildcarding or not.
	
	fmt.Println(URLParse("lower.case"))
	// "" <nil>
	// ^ We need to be able to allow this case since it is already a DNS1123 hostname.
	// We could do this by first checking if the given input is already a valid DNS1123 name ¯\_(ツ)_/¯
	
	fmt.Println(URLParse("lower.case:80"))
	// "" <nil>
	// ^ We need to allow this as well, similarly to above
	
	fmt.Println(URLParse("s3://lower.case"))
	// lower.case <nil>
	// ^ OK
	
	fmt.Println(URLParse("s3://ALL.CAPS"))
	// ALL.CAPS <nil>
	// ^ DNS1123 names MUST be lower case, so it would be a good idea for us to lower-case the output to handle capitals in the endpoint
	
	fmt.Println(URLParse("UPPER.CASE"))
	// "" <nil>
	// ^ We can probably allow this case since it is already a DNS1123 hostname [when lower-cased].
	// We could do this by first checking if the given input is already a valid DNS1123 name [when lower-cased] ¯\_(ツ)_/¯
```

@cbodley do you have thoughts about proto headers other than `http`/`https` (e.g., `s3://<endpoint>`? What about `<bucketname>.192.168.1.1` wildcarding?

Reply:
Update, found another problem:

```go
	fmt.Println(URLParse("https://http://my.host:80"))
	// http <nil>
	// ^ obviously wrong
```

Reply:
I've been thinking about this a little, and I think the underlying issue is that this is an "endpoint" and not a "URL." I suspect we might be able to find out what makes a valid "endpoint" as RGW/S3 is concerned and stick to that definition.

A URL, for example, can have fields like `?param=value` or `#subheader` or `mailto:my@email.com`, and all kinds of stuff that aren't really reasonable user inputs here. 

Practicallly, I think we can help ourselves out with a combination of CRD validations and some updates to the `URLParse()` code. We still need inputs from Casey and others in RGW team, but I think we can probably treat "endpoint" as a definition like this:

- An endpoint has this format: `[<protocol>://]<host>[:<port>]`
- `<protocol>` is optional and must be one of `http://` or `https://`
- `<host>` can be any valid DNS1123 hostname 
  - AND we support upper case letters here for convenience
- `<port>` is optional and must be in the form `<port-num>`
  - If a colon is present, `<port>` must also be present

Reply:
rgw's requirement for zone/zonegroup endpoints is that we can append a path and optional query string, then pass the result to libcurl via https://curl.se/libcurl/c/CURLOPT_URL.html. if the endpoint doesn't end with a / we add one with the path

the scheme part is optional according to the curl doc:
> If the given URL is missing a scheme name (such as "http://" or "ftp://" etc) then libcurl guesses based on the host. If the outermost subdomain name matches DICT, FTP, IMAP, LDAP, POP3 or SMTP then that protocol gets used, otherwise HTTP is used.

the presence of a path other than "/" would probably break multisite replication because rgw always hosts the s3 API at the root path

Reply:
Could you clarify what protocol headers that implies? I assume that because S3 uses HTTP, the only supported protocols would reasonably be `http://` and `https://` right? No one would put an S3 server behind `ftp://`, `ldap://`, `s3://` or `<random>://` would they?

Reply:
Also worth asking, does RGW support adding user info like user name/pass to the endpoint like `https://user:pass@host:443`?

Reply:
I think a function like this (https://play.golang.com/p/X6LmdKsp2WQ) would get us into a space where we can reliably parse things that I think people would reasonably enter into Rook.

It allows:
- omitting the protocol
- using nonstandard protocols since we don't have a solid answer as to whether only http/https are strictly valid for RGW (including `ftp://host`)
- using upper case letters in the endpoint host (for user convenience)
- IP address as hostname
- usage of `?param=val`, `#param=val` arguments if somehow needed to

It rejects:
- anything I can think of that can't parse to a valid hostname
- this: `"https://http://my.host:80",` (2 protos)
- this: `"http://my.host:80:8080",` (2 port nums)
- this: `"http://my.host:/",` (no port num)

@cbodley, could you take a detailed at the features I highlighted above to see if anything should be changed? You can also go to the play.golang.com link above and hit `Run` to check the outputs.

Reply:
FWIW, Kyle mentioned that `hdfs://<name node>:<port>/foo/bar` is a thing, and out of concern for whatever crazy stuff a user might be using as a scheme, I think we should allow anything as a scheme as a starting point. I'd rather not have Rook be in the business of keeping a list of supported RGW endpoint schemes unless we are certain that those are limited to `http`/`https`. If we find that users are shooting themselves in the feet with this scheme freedom, we can limit the scheme list at a later time.

Reply:
@thotz provided Casey doesn't flag any issues with what I proposed, please copy (but rename) the URLParse() function from my playground link above, and adapt all my `tests` as unit test inputs, along with any other tests that you can think of I might've missed. Please also make a note on the function that we are choosing to allow more scenarios for what is a valid endpoint than most users will probably use, with a link to here explaining more about why: https://github.com/rook/rook/pull/13326#discussion_r1490229603.